--- dhcp-0.10/dhcp6s.sh.orig	2005-09-18 22:48:31.000000000 +0200
+++ dhcp-0.10/dhcp6s.sh	2005-09-18 22:48:56.000000000 +0200
@@ -23,7 +23,7 @@
 [ ${NETWORKING} = "no" ] && exit 0
 
 # Check that files exist
-[ -f /usr/local/sbin/dhcp6s ] || exit 0
+[ -f /sbin/dhcp6s ] || exit 0
 [ -f /etc/dhcp6s.conf ] || exit 0
 [ -f /etc/server6_addr.conf ] || exit 0
 [ ${DHCP6SIF} = "" ] && exit 0
@@ -34,7 +34,7 @@
 start() {
 	# Start daemons.
 	echo -n $"Starting $prog: "
-	daemon /usr/local/sbin/dhcp6s -c /etc/dhcp6s.conf ${DHCP6SARGS} ${DHCP6SIF}
+	daemon /sbin/dhcp6s -c /etc/dhcp6s.conf ${DHCP6SARGS} ${DHCP6SIF}
 	RETVAL=$?
 	echo
 	[ $RETVAL -eq 0 ] && touch /var/lock/subsys/dhcp6s
--- dhcp-0.10/dhcp6c.sh~	2003-02-18 12:33:13.000000000 +0100
+++ dhcp-0.10/dhcp6c.sh	2005-09-18 22:50:43.000000000 +0200
@@ -22,7 +22,7 @@
 [ ${NETWORKING} = "no" ] && exit 0
 
 # Check that files exist
-[ -f /usr/local/sbin/dhcp6c ] || exit 0
+[ -f /sbin/dhcp6c ] || exit 0
 [ -f /etc/dhcp6c.conf ] || exit 0
 [ ${DHCP6CIF} = "" ] && exit 0
 
@@ -32,7 +32,7 @@
 start() {
 	# Start daemons.
 	echo -n $"Starting $prog: "
-	daemon /usr/local/sbin/dhcp6c -c /etc/dhcp6c.conf ${DHCP6CARGS} ${DHCP6CIF}
+	daemon /sbin/dhcp6c -c /etc/dhcp6c.conf ${DHCP6CARGS} ${DHCP6CIF}
 	RETVAL=$?
 	echo
 	[ $RETVAL -eq 0 ] && touch /var/lock/subsys/dhcp6c
@@ -52,7 +52,7 @@
 release() {
 	# Release addresses.
 	echo -n $"Releasing assigned addresses $prog: "
-	daemon /usr/local/sbin/dhcp6c ${DHCP6CARGS} -r all ${DHCP6CIF}
+	daemon /sbin/dhcp6c ${DHCP6CARGS} -r all ${DHCP6CIF}
 	RETVAL=$?
 	echo
 	return $RETVAL
