--- dhcp-0.10/common.h.no_strlcat	2007-02-16 15:30:18.000000000 -0500
+++ dhcp-0.10/common.h	2007-02-16 15:30:46.000000000 -0500
@@ -107,9 +107,6 @@
 extern struct host_conf *find_hostconf __P((const struct duid *));
 
 /* missing */
-#ifndef HAVE_STRLCAT
-extern size_t strlcat __P((char *, const char *, size_t));
-#endif
 #ifndef HAVE_STRLCPY
 extern size_t strlcpy __P((char *, const char *, size_t));
 #endif
--- dhcp-0.10/configure.in.no_strlcat	2007-02-16 15:30:18.000000000 -0500
+++ dhcp-0.10/configure.in	2007-02-16 15:30:40.000000000 -0500
@@ -67,7 +67,7 @@
 *)	;;
 esac
 dnl Checks for header files.
-AC_REPLACE_FUNCS(strlcpy strlcat)
+AC_REPLACE_FUNCS(strlcpy)
 AC_HEADER_STDC
 AC_CHECK_HEADERS(fcntl.h sys/ioctl.h sys/time.h syslog.h unistd.h ifaddrs.h)
 AC_MSG_CHECKING(for socklen_t)
