--- dhcp-0.10/Makefile.in.orig	2005-09-20 13:49:33.000000000 +0200
+++ dhcp-0.10/Makefile.in	2005-09-20 13:51:58.000000000 +0200
@@ -11,6 +11,7 @@
 YACC=	@YACC@
 LEX=	@LEX@
 TARGET=	dhcp6c dhcp6s
+CONFIGFILES= dhcp6c.conf dhcp6s.conf server6_addr.conf
 DESTDIR=
 
 INSTALL=@INSTALL@
@@ -97,6 +98,11 @@
 	$(INSTALL_DATA) dhcp6s.8 $(DESTDIR)$(mandir)/man8/
 	$(INSTALL_DATA) dhcp6c.conf.5 $(DESTDIR)$(mandir)/man5/
 	$(INSTALL_DATA) dhcp6s.conf.5 $(DESTDIR)$(mandir)/man5/
+	$(INSTALL) -d $(DESTDIR)$(sysconfigdir)
+	$(INSTALL) dhcp6s.sysconfig $(DESTDIR)$(sysconfigdir)/dhcp6s
+	$(INSTALL) dhcp6c.sysconfig $(DESTDIR)$(sysconfigdir)/dhcp6c
+	$(INSTALL) -d $(DESTDIR)$(etc)
+	$(INSTALL_PROGRAM) $(CONFIGFILES) $(DESTDIR)$(etc)
 
 rh_install:: install
 	$(INSTALL) -d $(initdir)
