--- dhcp-0.10/Makefile.in.orig	2005-09-18 22:04:44.000000000 +0200
+++ dhcp-0.10/Makefile.in	2005-09-18 22:06:38.000000000 +0200
@@ -91,17 +91,17 @@
 
 install::
 	$(INSTALL) -d $(DESTDIR)$(sbindir)
-	$(INSTALL_PROGRAM) -s -o bin -g bin $(TARGET) $(DESTDIR)$(sbindir)
+	$(INSTALL_PROGRAM) $(TARGET) $(DESTDIR)$(sbindir)
 	$(INSTALL) -d $(DESTDIR)$(mandir)/man8 $(DESTDIR)$(mandir)/man5
-	$(INSTALL_DATA) -o bin -g bin dhcp6c.8 $(DESTDIR)$(mandir)/man8/
-	$(INSTALL_DATA) -o bin -g bin dhcp6s.8 $(DESTDIR)$(mandir)/man8/
-	$(INSTALL_DATA) -o bin -g bin dhcp6c.conf.5 $(DESTDIR)$(mandir)/man5/
-	$(INSTALL_DATA) -o bin -g bin dhcp6s.conf.5 $(DESTDIR)$(mandir)/man5/
+	$(INSTALL_DATA) dhcp6c.8 $(DESTDIR)$(mandir)/man8/
+	$(INSTALL_DATA) dhcp6s.8 $(DESTDIR)$(mandir)/man8/
+	$(INSTALL_DATA) dhcp6c.conf.5 $(DESTDIR)$(mandir)/man5/
+	$(INSTALL_DATA) dhcp6s.conf.5 $(DESTDIR)$(mandir)/man5/
 
 rh_install:: install
 	$(INSTALL) -d $(initdir)
-	$(INSTALL_PROGRAM) -o root -g root dhcp6s.sh $(initdir)/dhcp6s
-	$(INSTALL_PROGRAM) -o root -g root dhcp6c.sh $(initdir)/dhcp6c
+	$(INSTALL_PROGRAM) dhcp6s.sh $(initdir)/dhcp6s
+	$(INSTALL_PROGRAM) dhcp6c.sh $(initdir)/dhcp6c
 	$(CHKCONFIG) --add dhcp6s
 	$(CHKCONFIG) --add dhcp6c
 
