--- dhcpv6-1.0.3/configure.ac~	2007-12-04 02:36:02.000000000 +0100
+++ dhcpv6-1.0.3/configure.ac	2007-12-14 01:19:26.896720416 +0100
@@ -45,6 +45,7 @@
 LT_RELEASE=$DHCPV6_MAJOR_VERSION.$DHCPV6_MINOR_VERSION
 LT_CURRENT=`expr $DHCPV6_MICRO_VERSION - $DHCPV6_INTERFACE_AGE`
 LT_REVISION=$DHCPV6_INTERFACE_AGE
+LT_AGE=0
 AC_SUBST(DHCPV6_VERSION)
 AC_SUBST(LT_RELEASE)
 AC_SUBST(LT_CURRENT)
--- dhcpv6-1.0.3/etc/sysconfig/Makefile.am~	2007-11-16 18:01:20.000000000 +0200
+++ dhcpv6-1.0.3/etc/sysconfig/Makefile.am	2007-12-15 20:07:12.682073227 +0200
@@ -8,6 +8,6 @@
 EXTRA_DIST = dhcp6s.sysconfig dhcp6r.sysconfig
 
 install-data-local:
-	$(MKDIR_P) $(DESTDIR)$(sysconfdir)/sysconfig
+	$(mkdir_p) $(DESTDIR)$(sysconfdir)/sysconfig
 	$(install_sh_DATA) dhcp6s.sysconfig $(DESTDIR)$(sysconfdir)/sysconfig/dhcp6s
 	$(install_sh_DATA) dhcp6r.sysconfig $(DESTDIR)$(sysconfdir)/sysconfig/dhcp6r
--- dhcpv6-1.0.3/etc/init.d/Makefile.am~	2007-11-16 19:58:09.000000000 +0200
+++ dhcpv6-1.0.3/etc/init.d/Makefile.am	2007-12-15 20:14:40.915809385 +0200
@@ -10,6 +10,6 @@
 rc_d_dir = $(DESTDIR)$(sysconfdir)/rc.d/init.d
 
 install-data-local:
-	$(MKDIR_P) $(rc_d_dir)
+	$(mkdir_p) $(rc_d_dir)
 	$(install_sh_SCRIPT) -m 0755 dhcp6s.sh $(rc_d_dir)/dhcp6s
 	$(install_sh_SCRIPT) -m 0755 dhcp6r.sh $(rc_d_dir)/dhcp6r
--- dhcpv6-1.0.3/etc/Makefile.am~	2007-11-16 20:06:21.000000000 +0200
+++ dhcpv6-1.0.3/etc/Makefile.am	2007-12-15 20:15:39.183155530 +0200
@@ -10,6 +10,6 @@
 MAINTAINERCLEANFILES = Makefile.in
 
 install-data-local:
-	$(MKDIR_P) $(DESTDIR)$(sysconfdir)
+	$(mkdir_p) $(DESTDIR)$(sysconfdir)
 	$(install_sh_DATA) dhcp6s.conf $(DESTDIR)$(sysconfdir)/dhcp6s.conf
 	$(install_sh_DATA) dhcp6c.conf $(DESTDIR)$(sysconfdir)/dhcp6c.conf
